pkg/api/profile.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"net/http"
+	"os"
+)
+
+const (
+	// ProfilePrefix is the URL prefix of profile APIs
+	ProfilePrefix = "/profile"
+	// StartAction is the URL for starting profiling
+	StartAction = "start"
+	// StopAction is the URL for stopping profiling
+	StopAction = "stop"
+)
+
+type (
+	// ProfileStatusResponse contains cpu and memory profile file paths
+	ProfileStatusResponse struct {
+		CPUPath    string `yaml:"cpuPath"`
+		MemoryPath string `yaml:"memoryPath"`
+	}
+
+	// StartProfilingRequest contains file path to profile file
+	StartProfilingRequest struct {
+		Path string `yaml:"path"`
+	}
+)
+
+func (s *Server) profileAPIEntries() []*Entry {
+	return []*Entry{
+		{
+			Path:    ProfilePrefix,
+			Method:  http.MethodGet,
+			Handler: s.getProfileStatus,
+		},
+		{
+			Path:    fmt.Sprintf("%s/%s/cpu", ProfilePrefix, StartAction),
+			Method:  http.MethodPost,
+			Handler: s.startCPUProfile,
+		},
+		{
+			Path:    fmt.Sprintf("%s/%s/memory", ProfilePrefix, StartAction),
+			Method:  http.MethodPost,
+			Handler: s.startMemoryProfile,
+		},
+		{
+			Path:    fmt.Sprintf("%s/%s", ProfilePrefix, StopAction),
+			Method:  http.MethodPost,
+			Handler: s.stopProfile,
+		},
+	}
+}
+
+func (s *Server) getProfileStatus(w http.ResponseWriter, r *http.Request) {
+	cpuFile := s.opt.CPUProfileFile
+	memFile := s.opt.MemoryProfileFile
+
+	result := &ProfileStatusResponse{CPUPath: cpuFile, MemoryPath: memFile}
+	w.Header().Set("Content-Type", "text/vnd.yaml")
+	err := yaml.NewEncoder(w).Encode(result)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (s *Server) startCPUProfile(w http.ResponseWriter, r *http.Request) {
+	spr := StartProfilingRequest{}
+	err := yaml.NewDecoder(r.Body).Decode(&spr)
+	if err != nil {
+		HandleAPIError(w, r, http.StatusBadRequest, fmt.Errorf("bad request"))
+		return
+	}
+
+	if spr.Path == "" || !IsPathValid(spr.Path) {
```

Comment:
I comment before to validate `spr.Path`, but I think we only need to check if it is an empty string.
because even `IsPathValid` returns true, it does not mean the path is valid when it is actually used.
and for L99, we can update `UpdateCPUProfile` to return an error and forward the error to the client.